std/algorithm/iteration.d

```diff
@@ -4280,3 +4285,307 @@ unittest
+    the median of r
+*/
+auto median(alias less = "a < b", R)(SortedRange!(R, less) r)
+if (isBidirectionalRange!R &&
+    !hasLength!R &&
+    !isRandomAccessRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
+}
+body
+{
+    uint length;
+    ElementType!R front_item;
+    ElementType!R back_item;
+
+    while (!r.empty)
+    {
+        front_item = r.front;
+        r.popFront;
+
+        back_item = r.back;
+        r.popBack;
+
+        ++length;
+    }
+
+    if (length % 2 == 0)
+    {
+        return (front_item + back_item) / 2;
+    }
+    else
+    {
+        return front_item;
+    }
+
+    assert(0);
+}
+
+/**
+Finds the median of a random access range. This function assumes
+r is an already sorted range. See $(D SortedRange) for details.
+This function is $(D Ο(1)).
+
+Params:
+    r = a random access range
+
+Returns:
+    the median of r
+*/
+auto median(alias less = "a < b", R)(SortedRange!(R, less) r)
+if (isRandomAccessRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
+}
+body
+{
+    if (r.length == 1)
+    {
+        return r[0];
+    }
+    else if (r.length == 2)
+    {
+        return (r[0] + r[1]) / 2;
+    }
+
+    if (r.length % 2 == 0)
+    {
+        return (r[($ / 2) - 1] + r[$ / 2]) / 2;
+    }
+    else
+    {
+        return r[$ / 2];
+    }
+}
+
+///
+@safe unittest
+{
+    import std.algorithm.sorting : sort;
+    import std.range : assumeSorted;
+    import std.math : approxEqual;
+
+    assert(sort([48, 31, 1, 37, 89, 83, 71]).median == 48);
+
+    assert([3, 5, 12].assumeSorted.median == 5);
+    assert([3, 5, 7, 13, 21, 23, 23, 39].assumeSorted.median == 17);
+    assert([2.9, 6.2, 6.5, 20.5].assumeSorted.median.approxEqual(6.35));
+    assert([2.9, 5.8, 6.2, 6.5, 20.5].assumeSorted.median.approxEqual(6.2));
+
+    assert([12, 5, 3].assumeSorted!"a > b".median == 5);
+}
+
+unittest
+{
+    import std.internal.test.dummyrange;
+    import std.range : assumeSorted;
+    import std.math : approxEqual;
+
+    auto r1 = new ReferenceForwardRange!int([3, 5, 12]);
+    assert(r1.assumeSorted.median == 5);
+
+    auto r2 = new ReferenceForwardRange!double([2.9, 6.2, 6.5, 20.5]);
+    assert(r2.assumeSorted.median.approxEqual(6.35));
+
+    auto r3 = new ReferenceForwardRange!int([12, 5, 3]);
+    assert(r3.assumeSorted!"a > b".median == 5);
+
+    DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Forward) r4;
+    assert(r4.assumeSorted.median == 5);
+
+    DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Forward) r5;
+    assert(r5.assumeSorted.median == 5);
+
+    DummyRange!(ReturnBy.Reference, Length.No, RangeType.Bidirectional) r6;
+    assert(r6.assumeSorted.median == 5);
+
+    DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Bidirectional) r7;
+    assert(r7.assumeSorted.median == 5);
+
+    DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Bidirectional) r8;
+    assert(r8.assumeSorted.median == 5);
```

Comment:
this could be a bit more DRY

```d
foreach(RV; TypeTuple!(ReturnBy.Reference,ReturnBy.Value)) {
    foreach(L; TypeTuple!(Length.No,Length.Yes)) {
    ....
    }
}
```